client/consensus/api/src/import_queue.rs

```diff
@@ -0,0 +1,304 @@
+	pub hash: <B as BlockT>::Hash,
+	/// Block header if requested.
+	pub header: Option<<B as BlockT>::Header>,
+	/// Block body if requested.
+	pub body: Option<Vec<<B as BlockT>::Extrinsic>>,
+	/// Justification(s) if requested.
+	pub justifications: Option<Justifications>,
+	/// The peer, we received this from
+	pub origin: Option<Origin>,
+	/// Allow importing the block skipping state verification if parent state is missing.
+	pub allow_missing_state: bool,
+	/// Skip block execution and state verification.
+	pub skip_execution: bool,
+	/// Re-validate existing block.
+	pub import_existing: bool,
+	/// Do not compute new state, but rather set it to the given set.
+	pub state: Option<ImportedState<B>>,
+}
+
+/// Verify a justification of a block
+#[async_trait::async_trait]
+pub trait Verifier<B: BlockT>: Send + Sync {
+	/// Verify the given data and return the BlockImportParams and an optional
+	/// new set of validators to import. If not, err with an Error-Message
+	/// presented to the User in the logs.
+	async fn verify(
+		&mut self,
+		origin: BlockOrigin,
+		header: B::Header,
+		justifications: Option<Justifications>,
+		body: Option<Vec<B::Extrinsic>>,
+	) -> Result<(BlockImportParams<B, ()>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String>;
+}
+
+/// Blocks import queue API.
+///
+/// The `import_*` methods can be called in order to send elements for the import queue to verify.
+/// Afterwards, call `poll_actions` to determine how to respond to these elements.
+pub trait ImportQueue<B: BlockT>: Send {
+	/// Import bunch of blocks.
+	fn import_blocks(&mut self, origin: BlockOrigin, blocks: Vec<IncomingBlock<B>>);
+	/// Import block justifications.
+	fn import_justifications(
+		&mut self,
+		who: Origin,
+		hash: B::Hash,
+		number: NumberFor<B>,
+		justifications: Justifications
+	);
+	/// Polls for actions to perform on the network.
+	///
+	/// This method should behave in a way similar to `Future::poll`. It can register the current
+	/// task and notify later when more actions are ready to be polled. To continue the comparison,
+	/// it is as if this method always returned `Poll::Pending`.
+	fn poll_actions(&mut self, cx: &mut futures::task::Context, link: &mut dyn Link<B>);
+}
+
+/// Hooks that the verification queue can use to influence the synchronization
+/// algorithm.
+pub trait Link<B: BlockT>: Send {
+	/// Batch of blocks imported, with or without error.
+	fn blocks_processed(
+		&mut self,
+		_imported: usize,
+		_count: usize,
+		_results: Vec<(BlockImportResult<B>, B::Hash)>
+	) {}
+
+	/// Justification import result.
+	fn justification_imported(
+		&mut self,
+		_who: Origin,
+		_hash: &B::Hash,
+		_number: NumberFor<B>,
+		_success: bool) {}
+
+	/// Request a justification for the given block.
+	fn request_justification(&mut self, _hash: &B::Hash, _number: NumberFor<B>) {}
+}
+
+/// Block import successful result.
+#[derive(Debug, PartialEq)]
+pub enum BlockImportStatus<N: std::fmt::Debug + PartialEq> {
```

Comment:
Fine with me, the only alternative I can come up with would be `BlockImportReceipt`.